### DISCUSSION TITLE
How to establish multiple IKE SAs between same peers.

### DISCUSSION THREAD

**Discussion by Jaditya-89** [strongswan/strongswan#2019] [General]
Hi, 

I have 2 hosts each of which have 2 NICs connected to each other. I need to establish the IPsec session between NICs1 and NICs2 of both hosts. The certificate I am using is based on the hostname which is same for NIC1 and NIC2 of host.

Auth Method: `pubkey`. 

The connection file is loaded programmatically with `start_action = start `in the conn file 
Initially, I set  `unique` value  to default. So, I am seeing duplicate sessions between the NICs . So,

2 IKE SAs between NIC1 Host1 <---> NIC1 Host2 
2 IKE SAs between NIC2 Host1 <---> NIC2 Host2 

Now, if I use `unique` value as replace, all above 4 SAs are getting replaced and a single IKE SA only exists between them, as the certificate is having identity based on host name. Reading the docs for strongswan, i understand that the uniqueness is checked on the remote id (which is equal to the hostname). Is it possible to have this uniqueness check based on the remote ID + IP of the remote peer.So, in this way we would have only 1 IKE SA between each NIC ?
Please let me know if more info is needed.
Thanks

**Comment by tobiasbrunner** [comment]
The whole point of the uniqueness check is to enforce that only a single IKE_SA exists between two identities if the IP address changes. However, duplicate IKE_SAs with different address families are allowed, so you could use IPv4 for one and IPv6 for the other. And you could always just disable the uniqueness check.

You could also consider using a single IKE_SA and MOBIKE to switch from one interface to the other if it goes down or the routing changes.

**Reply by Jaditya-89** [reply]
Okay, considering a scenario where I have more than 2 NICs attached, then using different address families won't work here. And having duplicate entries would cause redundant entries to be present in kernal SADB. If I am sure IP addresses won't be changing, is it possible to have a uniqueness check based on remote id and remote peer IP.
Basically, I have a given example of NIC here. Such scenarios can happen when we want  IPsec sessions between multiple entities which are present in a host machine where we use certificates of host and I want a unique SA between entities (use unique keyword)

**Reply by tobiasbrunner** [reply]
> And having duplicate entries would cause redundant entries to be present in kernal SADB.

What do you mean? What's your goal then?

> is it possible to have a uniqueness check based on remote id and remote peer IP.

What do you consider is the difference between this and disabling the unqieuness check?

> Basically, I have a given example of NIC here. Such scenarios can happen when we want IPsec sessions between multiple entities which are present in a host machine where we use certificates of host and I want a unique SA between entities (use unique keyword)

Could you please clarify what you mean exactly.

**Reply by Jaditya-89** [reply]
Okay, Let me rephrase my requirement.

I have two hosts where certain number of virtual machines are running. I need to setup IPSec in Transport mode between these vm's of both hosts. The IPs of the VM's are different here. The Auth method I am using is pubkey, and certificate is having SAN of the host name. 
The config I am using is having `start_action = start` Now, if i create IPSec sessions between them, VM's on both hosts might trigger IKE_SA_INIT between them, so due to this I am seeing duplicate SA entries ( basically 2) per vm pair. But I need to have only one entry, remove other redundant entries as only one will be used here. 
Now, if I use` unique = replace `keyword, as the remote id is based on the SAN field of the certificate which is unique per host, all SAs are getting deleted except one. 

So, my requirement is : Is there any way where I can have only single IKE SA entry per vm pair. 
May be modify the uniqueness check here to a combination of remote ID plus IP of remote peer. 

Thanks

**Reply by tobiasbrunner** [reply]
> But I need to have only one entry, remove other redundant entries as only one will be used here.

There is nothing you can do about this. If both sides can initiate SAs concurrently, there is no way to prevent all duplicates (the uniqueness check is prone to races). Those duplicates are no issue at all, though.

**Reply by Jaditya-89** [reply]
Understood, but just to understand more, won't this cause an impact on the scaling supported by strongswan ?

**Reply by tobiasbrunner** [reply]
What scaling are you referring to?
